packages/flutter/lib/src/material/material.dart

```diff
@@ -50,33 +125,376 @@ class Material extends StatelessComponent {
+    double radius;
+    if (containedInWell) {
+      radius = _getSplashTargetSize(referenceBox.size, position);
+    } else {
+      radius = _kDefaultSplashRadius;
+    }
+    _InkSplash splash = new _InkSplash(
+      renderer: this,
+      referenceBox: referenceBox,
+      position: position,
+      targetRadius: radius,
+      clipToReferenceBox: containedInWell,
+      onRemoved: onRemoved
+    );
+    _inkFeatures.add(splash);
+    inkChanged();
+    return splash;
+  }
+
+  double _getSplashTargetSize(Size bounds, Point position) {
+    double d1 = (position - bounds.topLeft(Point.origin)).distance;
+    double d2 = (position - bounds.topRight(Point.origin)).distance;
+    double d3 = (position - bounds.bottomLeft(Point.origin)).distance;
+    double d4 = (position - bounds.bottomRight(Point.origin)).distance;
+    return math.max(math.max(d1, d2), math.max(d3, d4)).ceilToDouble();
+  }
+
+  InkHighlight highlightRectAt({ RenderBox referenceBox, Color color, VoidCallback onRemoved }) {
+    _InkHighlight highlight = new _InkHighlight(
+      renderer: this,
+      referenceBox: referenceBox,
+      color: color,
+      onRemoved: onRemoved
     );
+    _inkFeatures.add(highlight);
+    inkChanged();
+    return highlight;
   }
+
+  void _removeFeature(_InkFeature feature) {
+    _inkFeatures.remove(feature);
+    inkChanged();
+  }
+
+  void inkChanged() {
+    _paintNeeded = true;
+    if (onPaintNeeded != null)
+      onPaintNeeded();
+    else
+      markNeedsPaint();
+  }
+
+  bool hitTestSelf(Point position) => true;
+
+  void paint(PaintingContext context, Offset offset) {
+    if (_inkFeatures.isNotEmpty) {
+      if (onPaintNeeded == null) {
+        final PaintingCanvas canvas = context.canvas;
+        canvas.save();
+        canvas.translate(offset.dx, offset.dy);
+        canvas.clipRect(Point.origin & size);
+        for (_InkFeature inkFeature in _inkFeatures)
+          inkFeature.paint(canvas);
+        canvas.restore();
+        _paintNeeded = false;
+      }
+    } else {
+      _paintNeeded = false;
+    }
+    assert(!_paintNeeded);
+    super.paint(context, offset);
+  }
+}
+
+class _InkFeatures extends OneChildRenderObjectWidget {
+  _InkFeatures({
+    Key key,
+    Widget child,
+    this.onPaintNeeded
+  }) : super(key: key, child: child);
+
+  final VoidCallback onPaintNeeded;
+
+  _RenderInkFeatures createRenderObject() => new _RenderInkFeatures(onPaintNeeded: onPaintNeeded);
+
+  void updateRenderObject(_RenderInkFeatures renderObject, _InkFeatures oldWidget) {
+    renderObject.onPaintNeeded = onPaintNeeded;
+  }
+}
+
+abstract class _InkFeature implements PaintableInkFeature {
+  _InkFeature({
+    this.renderer,
+    this.referenceBox,
+    this.onRemoved
+  });
+
+  final _RenderInkFeatures renderer;
+  final RenderBox referenceBox;
+  final VoidCallback onRemoved;
+
+  bool _debugExtant = true;
```

Comment:
I'd probably call this `_debugDisposed` and flip the bool.  That way it's clear there isn't another concept involved.